src/hotspot/share/runtime/os.cpp

```diff
@@ -1211,6 +1210,23 @@ bool os::is_first_C_frame(frame* fr) {
   if (old_fp - ufp > 64 * K) return true;
 
   return false;
+#endif
+}
+
+// Looks like all platforms can use the same function to check if C
+// stack is walkable beyond current frame.
+// This version of the method adds more checks that need the thread
+bool os::is_first_C_frame(frame* fr, Thread *t) {
+#ifdef _WINDOWS
+  return true; // native stack isn't walkable on windows this way.
+#else
+  return !fr->can_access_link(t) || os::is_first_C_frame(fr) ||
+         !t->is_in_full_stack((address)fr->sp()) ||
+         !t->is_in_full_stack((address)fr->fp()) ||
+         !t->is_in_full_stack((address)fr->sender_sp()) ||
+         !t->is_in_full_stack((address)fr->link());
```

Comment:
Isn't this check of `fr.link()` what you already did in `can_access_link`?

Reply:
You're correct.

Reply:
But as I said, I'm going to remove these checks all to gether.